compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2108,11 +2108,20 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 if let Some((expected, found)) = trace.values.ty() {
                     match (expected.kind(), found.kind()) {
                         (ty::Tuple(_), ty::Tuple(_)) => {}
-                        // If a tuple of length one was expected and the found expression has
-                        // parentheses around it, perhaps the user meant to write `(expr,)` to
-                        // build a tuple (issue #86100)
                         (ty::Tuple(fields), _) => {
-                            self.emit_tuple_wrap_err(&mut err, span, found, fields)
+                            if fields.len() == 0 {
```

Comment:
tiny nit: I would probably just make this a different pattern match case, like:

`(_, _) if expected.is_unit() =>`